gh-137530: generate an __annotate__ function for dataclasses __init__

DavidCEllis opened the pull request:
This is one approach to resolving #137530

It generates a new `__annotate__` function to attach to the `__init__` method and removes the in-line annotations that are now unused.

It's possible to keep the original source annotations if necessary, but they provide a second possibly incorrect source of the information and they will probably also cause additional issues with #135228 as they may keep the original class around when ForwardRef values are generated.

The generated `__annotate__` will also keep these references and need to be regenerated. I can extend this or follow up to handle these references but I didn't want to do too much in one PR.

<!-- gh-issue-number: gh-137530 -->
* Issue: gh-137530
<!-- /gh-issue-number -->

DavidCEllis committed:
replace source annotations in init with attached annotate function

DavidCEllis committed:
Include the new test for init annotations in all formats

DavidCEllis committed:
Add an extra check for init=False attributes

DavidCEllis committed:
Test that defining a forwardref makes init annotations resolve

DavidCEllis committed:
move _make_annotate_function to top level

needed for _add_slots

DavidCEllis committed:
stop unpacking a dictionary into another dictionary

JelleZijlstra commented on Lib/dataclasses.py:
```diff
@@ -524,6 +526,34 @@ def add_fns_to_class(self, cls):
+                    elif isinstance(v, annotationlib.ForwardRef):
+                        string_annos[k] = v.evaluate(format=annotationlib.Format.STRING)
+                    else:
+                        string_annos[k] = annotationlib.type_repr(v)
```
This is a bit unfortunate because we could get better STRING annotations by calling with the STRING format on the original class (and its base classes).

An approach to do this would be to make `__init__.__annotate__` delegate to calling `get_annotations()` on the class itself, and on any base classes that are contributing fields, with the appropriate format, then processing the result to add `"return": None` and filter out any fields that don't correspond to `__init__` parameters.

DavidCEllis replied:
I did think about that, but wasn't certain about the extra complexity, but I've since noticed that if there's a hint that *contains* a ForwardRef (like `list[undefined]`) the result leaks all the ForwardRef details.

I've added a test for that example to demonstrate why this is necessary.

I've taken a slightly different approach, in that I'm gathering all inherited annotations and using them to update the values from the fields. The original source annotation logic this replaces wasn't `__init__` specific so I've tried to keep the `__annotate__` generator non-specific too.

---

There are additional commits as this change meant that `__annotate__` now (always) has a reference to the original class which broke #135228 - this now also includes an additional test and fix for the issue I brought up there as I already needed to update the fields for this.

DavidCEllis replied:
I also unintentionally discovered that the `__annotate__` functions CPython generates for methods have an incorrect `__qualname__` while fixing the `__qualname__` for the function generated here.

DavidCEllis committed:
Get better 'Source' annotations - break slots gc tests

DavidCEllis committed:
Add an extra GC test for forwardrefs

DavidCEllis committed:
Remove references to original class from fields and annotate functions

DavidCEllis committed:
Test we don't fix annotations on hand-written init functions

DavidCEllis committed:
*Actually* test we don't fix annotations on hand-written init functions

The possibility to replace only occurs in _add_slots.

DavidCEllis committed:
Use the same annotation logic for VALUE, FORWARDREF and STRING

Also split out the single large test into multiple smaller tests

DavidCEllis committed:
Use a list of fields and return type instead of annotations dictionary.

Use `__class__` as the first argument name so `_update_func_cell_for__class__` can update it.

DavidCEllis committed:
remove outdated comment

JelleZijlstra commented on Lib/test/test_dataclasses/__init__.py:
```diff
@@ -2471,6 +2471,132 @@ def __init__(self, a):
+
+    def test_init_false_forwardref(self):
+        # Currently this raises a NameError even though the ForwardRef
+        # is not in the __init__ method
```
Add something like "This should ideally be fixed but we currently don't have a good way to fix it"?

JelleZijlstra commented on Lib/dataclasses.py:
```diff
@@ -524,6 +533,44 @@ def add_fns_to_class(self, cls):
+
+    # This is a flag for _add_slots to know it needs to regenerate this method
+    # In order to remove references to the original class when it is replaced
+    __annotate__._generated_by_dataclasses = True
```
Let's use a dunder name like `__generated_by_dataclasses__` since those are reserved for CPython; user code in theory could be using this attribute.

JelleZijlstra reviewed (commented):
Looks good, except for the small suggestions above.

DavidCEllis committed:
Include comment indicating the issue with VALUE annotations should ideally be resolved.

DavidCEllis committed:
Use a dunder name instead of a single underscore

DavidCEllis committed:
Use the dunder name in the tests too

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

encukou committed:
Add What's New entry

encukou committed:
Merge in the main branch
